File: app/src/main/java/com/hhh/paws/ui/petProfile/menu/procedures/ProceduresAdapter.java
package com.hhh.paws.ui.petProfile.menu.procedures;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.cardview.widget.CardView;
import androidx.recyclerview.widget.AsyncListDiffer;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;

import com.hhh.paws.R;
import com.hhh.paws.database.model.SurgicalProcedure;
import com.hhh.paws.ui.petProfile.menu.ItemClickListener;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class ProceduresAdapter extends RecyclerView.Adapter<ProceduresAdapter.ProceduresViewHolder> {

    static class ProceduresViewHolder extends RecyclerView.ViewHolder {

        public ProceduresViewHolder(@NonNull View itemView) {
            super(itemView);
        }

        CardView surgicalProceduresContainer = itemView.findViewById(R.id.surgicalProceduresContainer);
        TextView typeSurgicalProcedureItem = itemView.findViewById(R.id.typeSurgicalProcedureItem);
        TextView nameSurgicalProcedureItem = itemView.findViewById(R.id.nameSurgicalProcedureItem);
        TextView dateSurgicalProcedureItem = itemView.findViewById(R.id.dateSurgicalProcedureItem);
        TextView anesthesiaSurgicalProcedureItem = itemView.findViewById(R.id.anesthesiaSurgicalProcedureItem);
        TextView veterinarianSurgicalProcedureItem = itemView.findViewById(R.id.veterinarianSurgicalProcedureItem);

    }

    private List<SurgicalProcedure> oldList = new ArrayList<>();
    private final AsyncListDiffer<SurgicalProcedure> differ = new AsyncListDiffer<SurgicalProcedure>(
            this, callback
    );
    private static final DiffUtil.ItemCallback<SurgicalProcedure> callback = new DiffUtil.ItemCallback<SurgicalProcedure>() {
        @Override
        public boolean areItemsTheSame(@NonNull SurgicalProcedure oldItem, @NonNull SurgicalProcedure newItem) {
            return Objects.equals(oldItem.getId(), newItem.getId());
        }

        @Override
        public boolean areContentsTheSame(@NonNull SurgicalProcedure oldItem, @NonNull SurgicalProcedure newItem) {
            if (!Objects.equals(oldItem.getId(), newItem.getId())) {
                return false;
            } else if (!Objects.equals(oldItem.getAnesthesia(), newItem.getAnesthesia())) {
                return false;
            } else if (!Objects.equals(oldItem.getName(), newItem.getName())) {
                return false;
            } else if (!Objects.equals(oldItem.getDate(), newItem.getDate())) {
                return false;
            } else if (!Objects.equals(oldItem.getDescription(), newItem.getDescription())) {
                return false;
            } else if (!Objects.equals(oldItem.getType(), newItem.getType())) {
                return false;
            } else {
                return Objects.equals(oldItem.getVeterinarian(), newItem.getVeterinarian());
            }
        }
    };

    public void setDiffer(List<SurgicalProcedure> procedureList) {
        this.oldList.clear();
        this.oldList = procedureList;
        this.differ.submitList(oldList);
    }

    public void removeItemFromDiffer(SurgicalProcedure procedure) {
        this.oldList.remove(procedure);
        differ.submitList(oldList);
    }

    @NonNull
    @Override
    public ProceduresViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new ProceduresViewHolder(LayoutInflater.from(parent.getContext())
                .inflate(R.layout.procedure_item, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull ProceduresViewHolder holder, int position) {
        SurgicalProcedure procedure = differ.getCurrentList().get(position);
        holder.typeSurgicalProcedureItem.setText(procedure.getType());
        holder.anesthesiaSurgicalProcedureItem.setText(procedure.getAnesthesia());
        holder.nameSurgicalProcedureItem.setText(procedure.getName());
        holder.dateSurgicalProcedureItem.setText(procedure.getDate());
        holder.veterinarianSurgicalProcedureItem.setText(procedure.getVeterinarian());

        holder.itemView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                clickListener.onItemClickListener(differ.getCurrentList().get(holder.getAdapterPosition()));
            }
        });

        holder.itemView.setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View v) {
                clickListener.onItemLongClickListener(
                        differ.getCurrentList().get(holder.getAdapterPosition()),
                        holder.surgicalProceduresContainer
                );
                return false;
            }
        });
    }

    @Override
    public int getItemCount() {
        return differ.getCurrentList().size();
    }

    private ItemClickListener clickListener;
    public void setClickListener(ItemClickListener clickListener) {
        this.clickListener = clickListener;
    }
}

File: app/src/main/java/com/hhh/paws/ui/petProfile/menu/procedures/ProceduresAdapter.kt
package com.hhh.paws.ui.petProfile.menu.procedures

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.cardview.widget.CardView
import androidx.recyclerview.widget.AsyncListDiffer
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.hhh.paws.R
import com.hhh.paws.database.model.SurgicalProcedure
import com.hhh.paws.ui.petProfile.menu.ItemClickListener

class ProceduresAdapter: RecyclerView.Adapter<ProceduresAdapter.ProceduresViewHolder>() {

    inner class ProceduresViewHolder(itemView: View): RecyclerView.ViewHolder(itemView) {
        val surgicalProceduresContainer: CardView = itemView.findViewById(R.id.surgicalProceduresContainer)
        val typeSurgicalProcedureItem: TextView = itemView.findViewById(R.id.typeSurgicalProcedureItem)
        val nameSurgicalProcedureItem: TextView = itemView.findViewById(R.id.nameSurgicalProcedureItem)
        val dateSurgicalProcedureItem: TextView = itemView.findViewById(R.id.dateSurgicalProcedureItem)
        val anesthesiaSurgicalProcedureItem: TextView = itemView.findViewById(R.id.anesthesiaSurgicalProcedureItem)
        val veterinarianSurgicalProcedureItem: TextView = itemView.findViewById(R.id.veterinarianSurgicalProcedureItem)
    }

    private val callback: DiffUtil.ItemCallback<SurgicalProcedure> = object : DiffUtil.ItemCallback<SurgicalProcedure>() {
            override fun areItemsTheSame(oldItem: SurgicalProcedure, newItem: SurgicalProcedure): Boolean {
                return oldItem.id == newItem.id
            }

            override fun areContentsTheSame(oldItem: SurgicalProcedure, newItem: SurgicalProcedure): Boolean {
                return if (oldItem.id != newItem.id) {
                    false
                } else if (oldItem.anesthesia != newItem.anesthesia) {
                    false
                } else if (oldItem.name != newItem.name) {
                    false
                } else if (oldItem.date != newItem.date) {
                    false
                } else if (oldItem.description != newItem.description) {
                    false
                } else if (oldItem.type != newItem.type) {
                    false
                } else {
                    oldItem.veterinarian == newItem.veterinarian
                }
            }
        }

    private val differ = AsyncListDiffer(this, callback)

    fun setDiffer(procedureList: List<SurgicalProcedure>) {
        differ.submitList(procedureList)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ProceduresViewHolder {
        return ProceduresViewHolder(
            LayoutInflater.from(parent.context)
                .inflate(R.layout.procedure_item, parent, false)
        )
    }

    override fun onBindViewHolder(holder: ProceduresViewHolder, position: Int) {
        val procedure = differ.currentList[position]
        holder.typeSurgicalProcedureItem.text = procedure.type
        holder.anesthesiaSurgicalProcedureItem.text = procedure.anesthesia
        holder.nameSurgicalProcedureItem.text = procedure.name
        holder.dateSurgicalProcedureItem.text = procedure.date
        holder.veterinarianSurgicalProcedureItem.text = procedure.veterinarian
        holder.itemView.setOnClickListener { clickListener!!.onItemClickListener(differ.currentList[holder.adapterPosition]) }
        holder.itemView.setOnLongClickListener {
            clickListener!!.onItemLongClickListener(
                differ.currentList[holder.adapterPosition],
                holder.surgicalProceduresContainer
            )
            false
        }
    }

    override fun getItemCount(): Int {
        return differ.currentList.size
    }

    private var clickListener: ItemClickListener? = null
    fun setClickListener(clickListener: ItemClickListener?) {
        this.clickListener = clickListener
    }
}